README.rst
README
-------

The current processed web site with this Creative Commons
material is at
http://anh.cs.luc.edu/python/hands-on/3.1/handsonHtml

This is intended as about half of a CS0 course introducing ideas in Python.

It is intended as a very gentle introduction, just using the standard
Python distribution with its Idle IDE, and this tutorial with its examples folder.

A full CS1, major introduction through Python, is at https://runestone.academy.

changelog.rst
Hands-on Python Tutorial 3 Major Change Log
---------------------------------------------

July 29, 2019
~~~~~~~~~~~~~~~~~~

- Major review of the tutorial, creating version 2
- Added appendices for OS specific data, and for a debugging intro
- Removed lot of obsolete material on instalation, Idle, OS's.
- Elaborated a large number of explanations that had caused 
  problems in the past.
- Added paragraphs for ``for``, ``if``, ``while`` 
  on thinking of the English
  phrases that would suggest those compound statements.
- In the section on string formatting with **locals(),
  added the shorter alternate with f-strings.
- Have a new file to start Idle on Windows in the current folder.
- Did *not* change the numbering and names of sections and
  homework exercises (except replaced one subsection, 1.9.2,
  on a Python 3.3 bug, shifting some of the info on Idle editing).


June 5. 2014
~~~~~~~~~~~~~~~~~~

- Copied Windows Idle link file in examples to www folder.
- Clarified Play Computer Error Exercise

May 28. 2014
~~~~~~~~~~~~~~~~~~

Elaborated formatting explanation and issues in 
floating point arithmetic section, 1.14.3.


May 26. 2014
~~~~~~~~~~~~~~~~~~

Add example polygons.py - regular polygons and stars

May 18. 2014
~~~~~~~~~~~~~~~~~~

- Update for Python 3.4.
- Add Windows shortcut for Idle 3.4 

January 18. 2014
~~~~~~~~~~~~~~~~~~

Normalized picture sizes, particularly in pdf.

May 29. 2013
~~~~~~~~~~~~~

- Many clarifications and typo fixes in Ch 1
- Additions and clarifications to index

May 9, 2013
~~~~~~~~~~~

-  Added Section 1.1.6 Learning to Problem-Solve
-  Fixed garbled exercise 2.3.3.1

June 28, 2012
~~~~~~~~~~~~~

-  Added 3.1.6.4, Unique List Exercise

June 5, 2012
~~~~~~~~~~~~

-  Introduced the optional graphics idea of flushing updates on command.
   Two new example variants.
-  Added way to get videos from Dropbox - no ID needed.

May 23, 2012
~~~~~~~~~~~~

-  Fixed link and instructions for video downloads in 1.1.5.
   `<pythonOnMac.html>`_

May 20, 2012
~~~~~~~~~~~~

Many updates for Mac

-  References to updated document `Python On Mac <pythonOnMac.html>`_.
-  Added Mac info on using Idle and CGI scripts.
-  Fixed the Mac bug for the browseLocal function in Chapter 4,
   generated files do show automatically in the browser on a Mac.

January 26, 2012
~~~~~~~~~~~~~~~~

-  html version now lets you click the margin between left sidebar and
   the main text to toggle showing the sidebar.

January 14, 2012
~~~~~~~~~~~~~~~~

-  Updated some Mac instructions.
-  Changed the Mac/Unix script line for cgi files in examples/www
-  Replaced the graphics.py version, so it works on a Mac but it does
   not work from the Idle shell.

October 19, 2011
~~~~~~~~~~~~~~~~

-  In 4.4.4 added reminders of the final Python videos that do not
   exactlt follow the tutorial, but completely build examples.
-  Use tempname.html in browseLocal consistently in 4.3 and examples,
   rather than confusing special names each time.

October 6, 2011
~~~~~~~~~~~~~~~

-  Added www example CGIfix.py for Unix derived scripting situations
   with cgi files:
   Forces Unix newlines, adds script execution path if necesary, and
   makes the CGI script executable

September 19, 2011
~~~~~~~~~~~~~~~~~~

-  Added Text Triangle Exercise 2.4.12.2
-  Moved Graduate Exercise to 3.1.4 where it belongs.
-  Modified examples wages.py, wages1.py.

September 16, 2011
~~~~~~~~~~~~~~~~~~

-  Added missing column to table in example playComputerStub.rtf.
   Rewrote Play Computer excercises. Lots of typos fixed.

September 9, 2011
~~~~~~~~~~~~~~~~~

-  Added examples and exercises in 1.10, 1.11 on the string format
   method and functions returning a string formed with the string format
   method.

September 1, 2011
~~~~~~~~~~~~~~~~~

-  Elaborated example in 2.4.7.

August 25, 2011
~~~~~~~~~~~~~~~

-  Added a number of exercises in Chapter 3 for if-statements and
   while-statements, with a number of example stub files.
-  Added example files for cgi script template
-  Added links to videos in Google Docs in 1.1.5, including added videos
   for the end of Chapter 4, showing a full development of a simple web
   application.

June 23, 2011
~~~~~~~~~~~~~

-  Added example and information about tracebacks in 1.11.4.

June 13, 2011
~~~~~~~~~~~~~

-  Fixed bug! My example file types.py renamed types1.py (so does not
   conflict with system file types.py).
-  Added skeletonForWeb.py and skeletonFor.cgi to www examples.

June 6, 2011
~~~~~~~~~~~~

-  Elaborated explanations of dynamic web programming in 4.4, and added
   to the summary of CGI steps.

June 3, 2011
~~~~~~~~~~~~

-  Added simpler loop problems to 1.13.4, with accompanying example
   files.
-  Added data flow figure in chapter 4

May 25, 2011
~~~~~~~~~~~~

-  Lots of typo fixes while making videos in Chapter 3
-  Added example file matching 3.3.3 text: readLines0.py

May 20, 2011
~~~~~~~~~~~~

Brought explanation of backAndForth0.py animation up to date.

May 16, 2011
~~~~~~~~~~~~

-  Added documentation for additions to Zelle's graphics
-  Fixed source zip file and reference to it
-  Added link for iTunes U videos, now done for about half way through
   the tutorial.

May 10, 2011
~~~~~~~~~~~~

-  Note bug in command line operation of Windows version of 3.1 and 3.2,
   to be fixed in 3.2.1 later this month.

-  Patched madlib.py temporarily to fix the problem.

-  Added Mac specific instructions for donloading the files needed for
   Python and for dealing with files in Idle.

March 30, 2011
~~~~~~~~~~~~~~

-  Corrected display of return2.py in 1.11.6.

March 27, 2011
~~~~~~~~~~~~~~

-  Corrected formatBraces.py and the code copied into the tutorial.

March 10, 2011
~~~~~~~~~~~~~~

-  Added Python 3.2 references and Idle shortcuts in examples

March 1, 2011
~~~~~~~~~~~~~

-  bad formatting in 2.1.

-  typo in footnote order in 3.1

-  Separate Exercise Section for last Exercise 4.4.4.2

January 24, 2011
~~~~~~~~~~~~~~~~

-  few typos in io.html.


rstP/texFix.py
#!/usr/bin/env python3

import sys
startFar = 0 
startMissing = 0
endFar = 0

def fixTitlePage(s):
    #fix release, author

    with open('conf.py') as py:  # Python 2.6, not 3
        conf = py.read()

    start = "release = '"
    b = conf.index(start) + len(start)
    p = conf.index("'", b)
    ver = conf[b:p]
    print('ver =', ver)
    
    b = s.index(r'\release')
    p = s.index(r'\newcommand', b)
    print('>>>>>>>>>>old release\n' + s[b:p])

    s = (s[:b] +
         r'\release{' + ver + r''' for Python Version 3.1+}
\author{Dr. Andrew N. Harrington, Loyola University Chicago}
'''
         + s[p:])
    return s

def removeExtraContents(s):    
    # remove repeat of stuff in index page in html
    b = s.index('Python 3.1 Version')
    p = s.index(r'\chapter', b)
##    print('>>>>>>>>old index page stuff\n' + s[b:p])

    s = s[:b] +  s[p:]
    return s

def removeSearch(s):
    '''remove final ch: index and search'''
    b = s.rfind(r'\chapter')
    p = s.rfind(r'\renewcommand')
    print('ending junked\n' + s[b:p])

    s = s[:b] +  s[p:]
    return s


def main(name):
##    input('\n !!!!! Starting fix in Python, Press return to continue....:  ')
    with open(name) as f:
        s = f.read()
    s0 = s
    s = s.replace('↤', '$\\leftarrow$').replace('≠', '$\\neq$')
    s = s.replace('≤', '$\\leq$').replace('≥', '$\\geq$')
    # s = fixTitlePage(s)
    # s = removeExtraContents(s)
    
    # badline = r'\emph{genindex}'
    # s = s.replace(badline, '')
    # quoteindex = r'''\index{string!delimiters}\index{"}\index{'}'''
    # if quoteindex not in s:
    #     print(' ?? no quote index ??')
    # else:
    #     print('changing quote index')
    # input('Press return to continue...')
    # s = s.replace(quoteindex,
    #               r'''\index{string!delimiters}\index{"-quote}\index{'}''')
##    s = removeSearch(s)
    with open(name+'OLD', 'w') as f:
        f.write(s0)
    with open(name, 'w') as f:
        f.write(s)

if __name__ == '__main__':
    main(sys.argv[1])


rstP/fixtex.py
#!/usr/bin/env python3

def fixTex(name): 
    with open(name) as f:
        s = f.read()
    # fix non-ascii
    s =s.replace('↤', 'for a left arrow').replace(
        '≤', '$\\leq$').replace('≥', '$\\geq$').replace('≠', '$\\neq$')

    # fix title page
    #fix title
##    b = s.index(r'\title')
##    p = s.index(r'\date', b)
##    print('>>>>>>>>>>>>old title\n' + s[b:p])
##    s = (s[:b] +
##         r'\title{Hands-On Python Tutorial}' + '\n' +
##         s[p:])
##
##    #fix release, author
##
##    with open('conf.py') as py:  # Python 2.6, not 3
##        conf = py.read()
##
##    start = "release = '"
##    b = conf.index(start) + len(start)
##    p = conf.index("'", b)
##    ver = conf[b:p]
##    print('ver =', ver)
##    
##    b = s.index(r'\release')
##    p = s.index(r'\newcommand', b)
##    print('>>>>>>>>>>old release, author\n' + s[b:p])
##
##    s = (s[:b] +
##         r'\release{' + ver + r''' For Python 3.1}
##\author{Dr. Andrew N. Harrington\texttt{}~\\
##\texttt{Computer Science Department\\
##Loyola University Chicago}~\\
##{\footnotesize © Released under the Creative Commons 
##Attribution-Noncommercial-Share-Alike 3.0 
##United States License}
##{\footnotesize \url{http://creativecommons.org/licenses/by-nc-sa/3.0/us/}}}
##''' + s[p:])
##
    # --------------------
    
    # remove repeat of stuff in index page in html
    b = s.index('Python 3.1 Version')
    p = s.index(r'\chapter', b)
    print('>>>>>>>>old index page stuff\n' + s[b:p])

    s = s[:b] +  s[p:]

    # remove ch 5 index and search - WRONG if get index working
##    b = s.rfind(r'\chapter')
##    p = s.rfind(r'\end')
##    print('ending junked\n' + s[b:p])
##
##    s = s[:b] +  s[p:]
         
    with open(name, 'w') as f:
        f.write(s)
    print('Wrote tweaks to', name)

import sys\\
\hline
\end{tabulary}

if __name__ == '__main__':
    fixTex(sys.argv[1])
    
##sub = r'''
##\title{Hands-On Python Tutorial}
##\release{0.5 For Python 3.1}
##\author{Dr. Andrew N. Harrington\texttt{}~\\
##\texttt{Computer Science Department\\
##Loyola University Chicago}~\\
##{\footnotesize © Released under the Creative Commons 
##Attribution-Noncommercial-Share-Alike 3.0 
##United States License}
##{\footnotesize \url{http://creativecommons.org/licenses/by-nc-sa/3.0/us/}}}
##'''


TEXineq = '''
\begin{tabular}{|c|c|c|}
\hline
\textbf{Meaning} 
& 
\textbf{Math Symbol} 
& 
\textbf{Python Symbols}
\tabularnewline
\hline

Less than

 & 
\textless{}

 & 
\code{\textless{}}

\tabularnewline

Greater than

 & 
\textgreater{}

 & 
\code{\textgreater{}}

\tabularnewline

Less than or equal

 & 
$\leq$

 & 
\code{\textless{}=}

\tabularnewline

Greater than or equal

 & 
$\geq$

 & 
\code{\textgreater{}=}

\tabularnewline

Equals

 & 
=

 & 
\code{==}

\tabularnewline

Not equal

 & 
=$\neq$

 & 
\code{!=}

\tabularnewline
\hline
\end{tabular}


'''
